tokio/src/net/tcp/split_owned.rs

```diff
@@ -0,0 +1,224 @@
+//! `TcpStream` owned split support.
+//!
+//! A `TcpStream` can be split into an `OwnedReadHalf` and a
+//! `OwnedWriteHalf` with the `TcpStream::split_owned` method.
+//! `OwnedReadHalf` implements `AsyncRead` while `OwnedWriteHalf`
+//! implements `AsyncWrite`.
+//!
+//! Compared to the generic split of `AsyncRead + AsyncWrite`, this specialized
+//! split has no associated overhead and enforces all invariants at the type
+//! level.
+
+use crate::future::poll_fn;
+use crate::io::{AsyncRead, AsyncWrite};
+use crate::net::TcpStream;
+
+use bytes::Buf;
+use std::error::Error;
+use std::mem::MaybeUninit;
+use std::net::Shutdown;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{fmt, io};
+
+/// Owned read half of a [`TcpStream`], created by [`split_owned`].
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedReadHalf(Arc<TcpStream>);
+
+/// Owned write half of a [`TcpStream`], created by [`split_owned`].
+///
+/// Note that in the `AsyncWrite` implemenation of this type, `poll_shutdown` will
+/// shut down the TCP stream in the write direction.
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedWriteHalf(Arc<TcpStream>);
+
+pub(crate) fn split_owned(stream: TcpStream) -> (OwnedReadHalf, OwnedWriteHalf) {
+    let arc = Arc::new(stream);
```

Comment:
I actually thought it might work without the `Arc`, since the socket already might have heap-allocated state, and so we would only need to have an extra refcount in there instead of requiring another allocation. it seems like we might - there is a weak pointer somewhere.  But it's a few layers deeper, so that would be a bigger change.